docs/contribute/testing.adoc

```diff
@@ -0,0 +1,423 @@
+
+
+==== test_exit function
+NOTE: This function should only be called from the entry script which is
+responsible for running all the tests from a specific directory.
+
+The function is responsible for cleaning-up the testing environment. You can
+call it without arguments or with one argument -- a script/function which will
+do additional clean-up tasks.
+[[app-listing]]
+[source,bash]
+----
+test_exit [CLEAN_SCRIPT|CLEAN_FUNCTION]
+----
+
+
+==== require function
+Checks if requirements are in the `$PATH`, use it as follows:
+[[app-listing]]
+[source,bash]
+----
+require 'program' || return 255
+----
+
+
+==== probecheck function
+Checks if probe exists, use it as follows:
+[[app-listing]]
+[source,bash]
+----
+probecheck 'probe' || return 255
+----
+
+
+==== verify_results function
+Verifies that there is the `COUNT` number of results of selected OVAL `TYPE` in
+a `RESULTS_FILE`:
+[[app-listing]]
+[source,bash]
+----
+verify_results TYPE CONTENT_FILE RESULTS_FILE COUNT
+----
+
+NOTE: This function expects that the OVAL `TYPE` is numbered from `1` to `COUNT`
+in the `RESULTS_FILE`.
+
+
+==== assert_exists function
+Does an XPath query to a file specified in `$result` variable and checks if
+number of results matches with an expected number specified as an argument:
+[[app-listing]]
+[source,bash]
+----
+result="relative_path_to_file"
+assert_exists EXPECTED_NUMBER_OF_RESULTS XPATH_QUERY_STRING
+----
+
+For example, let's say you want to check that in the `results.xml` file the
+result of the rule `xccdf_com.example.www_rule_test` is fail:
+[[app-listing]]
+[source,bash]
+----
+result="./results.xml"
+my_rule_="xccdf_com.example.www_rule_test"
+assert_exists 1 "//rule-result[@idref=\"$my_rule\"]/result[text()=\"fail\"]"
+----
+
+
+=== Adding test files
+Now, as we know where a new test should go and what functions and capabilities
+are provided by the common test library, we can add test files which will
+contain test scripts and content required for testing our pull request.
+
+To sum up, we are adding tests to check the functionality of a single rule
+evaluation of the `xccdf` module of the `oscap` command line tool (see the commit
+link:https://github.com/OpenSCAP/openscap/commit/4637b132e710d7ab73601718a531a8587b406cd3[4637b13]
+for reference) and we have decided that it will go into the
+link:../../tests/API/XCCDF/unittests[tests/API/XCCDF/unittests] directory.
+
+We will add the `test_single_rule.sh` script which will contain our tests and
+multiple data files. We will also add multiple XML files which will contain:
+
+* two rules: a rule which always passes, and a rule which passes or fails based
+  on its value;
+* two profiles: a profile which has both rules selected and a profile which
+  has only one rule selected;
+* and a tailoring file which customizes one profile and selects only one rule.
+
+The `test_single_rule.sh` script will contain the following tests to check
+the functionality of the single rule evaluation feature:
+
```

Comment:
Idea: Does some clever book exist that explains how to create the test cases?

Reply:
This might be an overkill for testing guide, IMO most people would not read a book just to write a simple test for their code.